### ISSUE TITLE
Example of extending the UserProfile

### ISSUE THREAD

**Issue by mjmare** [wagtail/wagtail#12274]
There is documentation on how to integrate a custom user model. 
It would be nice to have documentation on extending the wagtail user profile. This would be convenient if one does not want to have extra field on the user model, but on the profile instead, as is usually recommended in Django country.
There already is documentation on changing the account settings (https://docs.wagtail.org/en/stable/extending/custom_account_settings.html#operating-on-the-userprofile-model) but there is no mention that the user profile should be extended, and how to do that. 

TIA

**Comment by gasman** [comment]
I don't think customising the user profile model is something we'd want to support - unlike the User model (where Wagtail has to work with whatever the developer has configured for the project) the UserProfile exists for Wagtail's own purposes. If a developer of a site or a Wagtail add-on wants to add their own user properties, they'd be better off doing that with their own model. (As [the relevant Django docs](https://docs.djangoproject.com/en/5.1/topics/auth/customizing/#extending-the-existing-user-model) say, there's nothing special about a user profile model - it's just an ordinary model with a one-to-one relation to User, and thus there's no reason why you can't have multiple ones for different apps.)

So, perhaps the real feature request here is to make it possible to update arbitrary additional related models from the account settings form, rather than form_object being limited to just `user` and `profile`.

**Comment by mjmare** [comment]
@gasman That makes sense. So indeed, additional user profiles would be useful.

**Comment by jhrr** [comment]
@gasman Sorry to mildly necrobump this thread, but there is one issue with cutomizing the Wagtail `UserProfile` that we're bumping into, and that is being able to replace or alias the `avatar` field on the `UserProfile` to some other avatar on some other model so it could use that instead. 

For example, we have a custom `User` model and a `Profile` model, and we keep the avatar on the `Profile`. We'd prefer to just be able to use the avatar from the `Profile` everywhere, across all sub-domains/contexts, including inside the CMS itself. Is there any way we can achieve this currently?

Perhaps we could add some kind of aliasing descriptor property to the Wagtail `UserProfile` model that would do a lookup on some other model and field combination specified in the `settings` if that setting was defined, otherwise just fall back to using `UserProfile.avatar`?

**Comment by gasman** [comment]
@jhrr Rather than trying to patch the `UserProfile` model to return something other than its own `avatar` property, I think it would make more sense to make the view-level code in wagtail.admin customisable, so that we can drop in some alternative code to tell it to retrieve the avatar from somewhere other than the Wagtail user profile.

Currently, any time we render an avatar in the admin, it goes through:
* the `{% avatar %}` tag in `wagtail/admin/templatetags/wagtailadmin_tags.py`
* which renders the template `wagtail/admin/templates/wagtailadmin/shared/avatar.html`
* which calls the `{% avatar_url %}` tag in `wagtail/admin/templatetags/wagtailadmin_tags.py`
* which accesses `user.wagtail_userprofile.avatar.url`

In principle, it's already possible to intercept this process by overriding the `wagtailadmin/shared/avatar.html` template within one of your own apps, but that would be a lot of brittle copy-and-paste, so I wouldn't really recommend that. Instead, we could introduce a [hook](https://docs.wagtail.org/en/stable/reference/hooks.html) within the `avatar_url` tag implementation, which would allow you to write something like this in your wagtail_hooks.py:

```python
@hooks.register("get_avatar_url")
def myapp_avatar(user, size):
    return user.myapp_profile.avatar.url
```

and then `avatar_url` would call any `get_avatar_url` hooks that exist, returning the result of the first one that returns a non-empty response, and falling back on the current logic if none do.

**Comment by jhrr** [comment]
> [@jhrr](https://github.com/jhrr) Rather than trying to patch the `UserProfile` model to return something other than its own `avatar` property, I think it would make more sense to make the view-level code in wagtail.admin customisable, so that we can drop in some alternative code to tell it to retrieve the avatar from somewhere other than the Wagtail user profile.
> 
> Currently, any time we render an avatar in the admin, it goes through:
> 
>     * the `{% avatar %}` tag in `wagtail/admin/templatetags/wagtailadmin_tags.py`
> 
>     * which renders the template `wagtail/admin/templates/wagtailadmin/shared/avatar.html`
> 
>     * which calls the `{% avatar_url %}` tag in `wagtail/admin/templatetags/wagtailadmin_tags.py`
> 
>     * which accesses `user.wagtail_userprofile.avatar.url`
> 
> 
> In principle, it's already possible to intercept this process by overriding the `wagtailadmin/shared/avatar.html` template within one of your own apps, but that would be a lot of brittle copy-and-paste, so I wouldn't really recommend that. Instead, we could introduce a [hook](https://docs.wagtail.org/en/stable/reference/hooks.html) within the `avatar_url` tag implementation, which would allow you to write something like this in your wagtail_hooks.py:
> 
> @hooks.register("get_avatar_url")
> def myapp_avatar(user, size):
>     return user.myapp_profile.avatar.url
> 
> and then `avatar_url` would call any `get_avatar_url` hooks that exist, returning the result of the first one that returns a non-empty response, and falling back on the current logic if none do.

Agreed this is a much more ergonomic and less invasive solution. Makes good sense to me.

If I'm reading you right this hook doesn't exist yet correct? If you can point me in the right direction (so I don't have to do too much spelunking, I'm under a load of work pressure, but I do need to work on some avatars that will involve this) I can potentially attempt to work up a patch and submit it...

**Comment by laymonage** [comment]
@gasman I wonder whether it would be better if we make use of Django's template tag `Library`'s mechanisms as the primary way to override Wagtail's template tags. It works as a registry, so currently it's already possible to do something like:

```py
from wagtail.admin.templatetags.wagtailadmin_tags import (
    avatar_url as wagtail_avatar_url,
)
from wagtail.admin.templatetags.wagtailadmin_tags import register as wagtail_templatetags


@wagtail_templatetags.simple_tag
def avatar_url(user, size=50, gravatar_only=False):
    if user.is_superuser:
        return "https://placehold.co/200x200"
    return wagtail_avatar_url(user, size, gravatar_only)
```

in a file inside one of your apps' `templatetags` directory. You can even do it in your `wagtail_hooks.py` if you wish.

The only problem is that this doesn't work in places where Wagtail imports the `avatar_url` function directly from `wagtailadmin_tags` and uses it as a function instead of a template tag. There's only one such case I found, which is in the code for the Account settings menu (the bottom left user avatar):

https://github.com/wagtail/wagtail/blob/c2676af857a41440e05e03038d85a540dcca3ce2/wagtail/admin/ui/sidebar.py#L263-L287

Unfortunately changing the import in that file to something like

```py
from wagtail.admin.templatetags.wagtailadmin_tags import register
avatar_url = register.tags.get("avatar_url")
```

doesn't work because the function registered in the template tag registry is the template tag version (that takes a parser and the token as arguments). A workaround is to refactor Wagtail's code so it does something like `render_to_string("{% avatar_url user size=50 %}", {"user": self.user})`.

I guess I kind of answered my own question here – implementing a hook won't require us changing the menu code, so it is simpler to get done. But I still wonder whether that refactoring might be worth it if we do want to go down this route for template tag customisations 🙂

**Comment by gasman** [comment]
@laymonage Ah, good catch! I failed to spot that one place that was bypassing the expected `{% avatar %}` -> `avatar.html` -> `{% avatar_url %}` pipeline and calling `avatar_url` as a function instead, because I was searching the codebase for `% avatar_url` ...

Even if it wasn't for that special case, I'd still be happier implementing our own extension point rather than leaning on the template tag registry, so that the implementation detail of using a template tag here doesn't become part of our public API and lock us in to using a template tag for ever more. (Come to think of it - why _are_ we using a tag for `{% avatar_url %}`? Why isn't it just a function that gets called from the `{% avatar %}` Python code, dropped into the template context and made available to `avatar.html` as a variable?)

@jhrr Yep, this hook doesn't exist yet, and if you're up for working on a patch, that would be much appreciated! The change would be [to this function](https://github.com/wagtail/wagtail/blob/23275a4cef4d36bb311abe315eb9ddfc43868e8b/wagtail/admin/templatetags/wagtailadmin_tags.py#L652-L659), and would look something like (untested):

```python
@register.simple_tag
def avatar_url(user, size=50, gravatar_only=False):
    """
    A template tag that receives a user and size and return
    the appropriate avatar url for that user.
    Example usage: {% avatar_url request.user 50 %}
    """
    for hook_fn in hooks.get_hooks("get_avatar_url"):
        url = hook_fn(user, size)
        if url:
            return url

    # followed by the existing code
```

This would need to be accompanied by unit tests - I'd suggest adding a hook function in `wagtail/test/testapp/wagtail_hooks.py` that returns a custom URL when some specific condition is true, such as the username containing "fred", and adding tests to confirm that both the custom code and the default avatar logic come into play at the appropriate times. The hook will also need to be documented in the [hooks reference docs](https://github.com/wagtail/wagtail/blob/main/docs/reference/hooks.md), and ideally we'd mention it in the [customising admin templates](https://github.com/wagtail/wagtail/blob/main/docs/advanced_topics/customization/admin_templates.md) docs too - this _could_ be a detailed howto, but I expect we can come up with a simple self-contained example snippet on the hooks reference page that's not too long, and just link to that.

**Comment by gasman** [comment]
I realise we've drifted off the original topic of this issue a bit, so I'll open a new one specifically about customising avatars.
